File: java/warmup-2/StringTimes.java
// Файл: java/warmup-2/StringTimes.java

public class StringTimes {

    /*
     --- Условие ---
     Given a string and a non-negative int n, return a larger string that is
     n copies of the original string.

     --- Мой анализ ---
     1. Понимание: Нужно повторить строку n раз. В Java нет оператора `*` для
        строк, поэтому требуется итеративный подход.
     2. Крайние случаи: n = 0. Цикл `for (int i = 0; i < 0; i++)` не
        выполнится ни разу, и метод вернет пустую строку, что корректно.
     3. Сложность:
        - Временная: O(N * M), где N=str.length(), M=n.
        - Пространственная: O(N * M).
     4. Альтернативы:
        - Простая конкатенация с `+` в цикле: менее эффективна из-за создания
          множества промежуточных объектов строк.
        - `str.repeat(n)`: доступно с Java 11, самый лаконичный способ.
        - `StringBuilder` (используется в решении): самый эффективный и
          общепринятый способ для построения строк в циклах.
    */

    // --- Решение ---
    public String stringTimes(String str, int n) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(str);
        }
        return sb.toString();
    }

    // --- Метод для самопроверки ---
    public static void main(String[] args) {
        StringTimes solution = new StringTimes();

        System.out.println("Тестирование метода stringTimes:");
        // Ожидаем: HiHi
        System.out.println(solution.stringTimes("Hi", 2));
        // Ожидаем: HiHiHi
        System.out.println(solution.stringTimes("Hi", 3));
        // Ожидаем: (пустая строка)
        System.out.println(solution.stringTimes("Hi", 0));
    }
}

File: python/warmup-2/string_times.py
# Файл: python/warmup-2/string_times.py

# --- Условие ---
# Given a string and a non-negative int n, return a larger string that is
# n copies of the original string.
#
# string_times('Hi', 2) → 'HiHi'
# string_times('Hi', 3) → 'HiHiHi'

# --- Мой анализ ---
# 1. Понимание: Повторить данную строку n раз.
# 2. Крайние случаи: n = 0. `str * 0` в Python корректно вернет
#    пустую строку `''`. n = 1 вернет саму строку.
# 3. Сложность:
#    - Временная: O(N * M), где N=len(str), M=n.
#    - Пространственная: O(N * M).
# 4. Альтернативы: Можно было бы использовать цикл for и конкатенацию,
#    но это было бы медленнее и многословнее, чем оператор `*`.

# --- Решение ---


def string_times(str, n):
    return str * n
